### PR TITLE
Add the basics of a test framework

### PR DESCRIPTION
**Pull request by Skarlso**

### CODE REVIEW

**Comment by Skarlso** [comment]
Slight modification from the previous merge strategy. This is now using Kubernetes and Kustomize merge tactics.

To define some defaults, you can directly influence the result by applying a partial merge to the object if only a small change is required. Anything more complex can be performed by hand after the default object is created.

A patch request can look something like this:

```go
	cv, err := env.CreateComponentVersion(WithComponentVersionPatch([]byte(`spec:
  repository:
    secretRef:
      name: test-overwrite
`)))
```

The long-winded reason for not using a struct is, in fact, an age-old issue of generating default values even for things that are marked `omitempty`. There is a 7 or so years old issue about it in Go -> https://github.com/golang/go/issues/11939

In short, you can't generate the above struct unless you define some custom MarshalJSON methods, and even then, how do you tell if a `0` is an actual set value or an omitempty generated default. Since it's not a pointer.

**Comment by Skarlso** [comment]
## Update

I got rid of the whole merge strategy thing because, thinking about it, it made no sense in the end.

If you want a change, you can just set it yourself in the test starting from the default like this:

```go
			obj := DefaultComponent.DeepCopy()
			obj.Spec.Version.Semver = tt.givenVersion
			obj.Status.ReconciledVersion = tt.reconciledVersion
```

This is a heck of a lot cleaner, with much less boilerplate.

**Review by souleb — APPROVED** [review]
LGTM!
